saleor/core/telemetry/utils.py

```diff
@@ -0,0 +1,139 @@
+import logging
+from collections.abc import Callable, Sequence
+from contextlib import contextmanager
+from contextvars import ContextVar
+from dataclasses import dataclass, field
+from enum import Enum
+from functools import wraps
+
+from opentelemetry.trace import Link, SpanContext, TraceFlags
+from opentelemetry.util.types import Attributes, AttributeValue
+
+logger = logging.getLogger(__name__)
+
+Amount = int | float
+
+_GLOBAL_ATTRS: ContextVar[dict[str, AttributeValue]] = ContextVar("global_attrs")
+
+
+class Scope(Enum):
+    CORE = "saleor.core"
+    SERVICE = "saleor.service"
+
+    @property
+    def is_service(self):
+        return self == Scope.SERVICE
+
+
+class Unit(Enum):
+    SECOND = "s"
+    MILLISECOND = "ms"
+    NANOSECOND = "ns"
+    REQUEST = "{request}"
+
+
+UNIT_CONVERSIONS: dict[tuple[Unit, Unit], float] = {
+    (Unit.NANOSECOND, Unit.MILLISECOND): 1e-6,
+    (Unit.NANOSECOND, Unit.SECOND): 1e-9,
+}
+
+
+def convert_unit(amount: Amount, unit: Unit | None, to_unit: Unit) -> Amount:
+    if unit is None or unit == to_unit:
+        return amount
+    try:
+        return amount * UNIT_CONVERSIONS[(unit, to_unit)]
+    except KeyError as e:
+        raise ValueError(f"Conversion from {unit} to {to_unit} not supported") from e
+
+
+@contextmanager
+def set_global_attributes(attributes: dict[str, AttributeValue]):
+    token = _GLOBAL_ATTRS.set(attributes)
+    try:
+        yield
+    finally:
+        _GLOBAL_ATTRS.reset(token)
+
+
+def get_global_attributes() -> dict[str, AttributeValue]:
+    return _GLOBAL_ATTRS.get({})
+
+
+def enrich_with_global_attributes(attributes: Attributes) -> Attributes:
+    return {**(attributes or {}), **get_global_attributes()}
+
+
+@dataclass(frozen=True)
+class TelemetryTaskContext:
+    """Carries telemetry context when propagated to Celery tasks."""
+
+    # TODO add TraceState support
```

Comment:
Do you have task for that? Maybe worth linking the issue number here

Reply:
We're tracking this in Linear, for now I'll skip linking the task here and we can fix it later.